### PR TITLE
replace servo with spectator in the client

### PR DESCRIPTION
**Pull request by joshgord**

### CODE REVIEW

File: eureka-core/src/test/java/com/netflix/eureka/registry/TimeConsumingInstanceRegistryTest.java [hunk 1]
```diff
@@ -67,6 +69,7 @@ public class TimeConsumingInstanceRegistryTest extends AbstractTester {
      * Note that there is a thread retrieving and printing out registry status for debugging purpose.
      */
     @Test
+    @Ignore("NumOfRenewsPerMinThreshold should be updated to 256")
```
Comment (joshgord) [inline eureka-core/src/test/java/com/netflix/eureka/registry/TimeConsumingInstanceRegistryTest.java:72]: Was already failing in the master branch

File: eureka-client-jersey2/build.gradle [hunk 2]
```diff
@@ -27,6 +27,8 @@ dependencies {
     compile project(':eureka-client')
     compile 'org.glassfish.jersey.core:jersey-client:2.23.1'
     compile 'org.glassfish.jersey.connectors:jersey-apache-connector:2.23.1'
+    compile "org.slf4j:slf4j-api:1.7.36"
```
Comment (joshgord) [inline eureka-client-jersey2/build.gradle:30]: slf4j apparently pulled in transitively via servo

File: eureka-client/src/main/java/com/netflix/discovery/shared/NamedConnectionPool.java [hunk 3]
```diff
@@ -76,15 +76,17 @@ public NamedConnectionPool(String name, ClientConnectionOperator operator,
...
+        final CompositeRegistry registry = Spectator.globalRegistry();
+        freeEntryCounter = registry.counter(name + "_Reuse");
+        createEntryCounter = registry.counter(name + "_CreateNew");
+        requestCounter = registry.counter(name + "_Request");
+        releaseCounter = registry.counter(name + "_Release");
+        deleteCounter = registry.counter(name + "_Delete");
+        requestTimer = registry.timer(name + "_RequestConnectionTimer");
+        creationTimer = registry.timer(name + "_CreateConnectionTimer");
+        PolledMeter.using(registry).withName("connectionCount")
+            .monitorValue(this.getConnectionsInPool());
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/NamedConnectionPool.java:88]: See note on using lambda

File: eureka-client/src/main/java/com/netflix/discovery/shared/NamedConnectionPool.java [hunk 4]
```diff
@@ -140,35 +142,27 @@ protected void deleteEntry(BasicPoolEntry entry) {
...
-        return releaseCounter.getValue().longValue();
+        return releaseCounter.count();
     }
 
     public final long getDeleteCount() {
-        return deleteCounter.getValue().longValue();
-    }
-
-    @Monitor(name = "connectionCount", type = DataSourceType.GAUGE)
-    public int getConnectionCount() {
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/NamedConnectionPool.java:163]: This is a public method, should it be deleted?

File: eureka-client/src/main/java/com/netflix/discovery/shared/resolver/AsyncResolver.java [hunk 5]
```diff
@@ -107,6 +107,11 @@ public AsyncResolver(String name,
         this.refreshIntervalMs = refreshIntervalMs;
         this.warmUpTimeoutMs = warmUpTimeoutMs;
 
+        final CompositeRegistry registry = Spectator.globalRegistry();
+        PolledMeter.using(registry)
+            .withName(METRIC_RESOLVER_PREFIX + "lastLoadTimestamp")
+            .monitorValue(getLastLoadTimestamp());
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/resolver/AsyncResolver.java:113]: See note on using lambda

File: eureka-client/src/main/java/com/netflix/discovery/shared/resolver/AsyncResolver.java [hunk 6]
```diff
@@ -133,14 +138,13 @@ public AsyncResolver(String name,
         );
 
         this.resultsRef = new AtomicReference<>(initialValue);
-        Monitors.registerObject(name, this);
+        PolledMeter.using(registry)
+            .withName(METRIC_RESOLVER_PREFIX + "endpointsSize")
+            .monitorValue(getEndpointsSize());
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/resolver/AsyncResolver.java:143]: See note on using lambda

File: eureka-client/src/main/java/com/netflix/discovery/shared/resolver/ReloadingClusterResolver.java [hunk 7]
```diff
@@ -74,12 +73,9 @@ public ReloadingClusterResolver(final ClusterResolverFactory<T> factory, final l
...
         }
-
-        try {
-            Monitors.registerObject(this);
-        } catch (Throwable e) {
-            logger.warn("Cannot register metrics", e);
-        }
+        PolledMeter.using(Spectator.globalRegistry())
+            .withName(METRIC_RESOLVER_PREFIX + "lastReloadTimestamp")
+            .monitorValue(getLastReloadTimestamp());
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/resolver/ReloadingClusterResolver.java:78]: See note on using lambda

File: eureka-client/src/main/java/com/netflix/discovery/shared/transport/decorator/SessionedEurekaHttpClient.java [hunk 8]
```diff
@@ -56,7 +55,9 @@ public SessionedEurekaHttpClient(String name, EurekaHttpClientFactory clientFact
         this.clientFactory = clientFactory;
         this.sessionDurationMs = sessionDurationMs;
         this.currentSessionDurationMs = randomizeSessionDuration(sessionDurationMs);
-        Monitors.registerObject(name, this);
+        PolledMeter.using(Spectator.globalRegistry())
+            .withName(METRIC_TRANSPORT_PREFIX + "currentSessionDuration")
+            .monitorValue(getCurrentSessionDuration());
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/transport/decorator/SessionedEurekaHttpClient.java:60]: See note on using lambda

File: eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/ApacheHttpClientConnectionCleaner.java [hunk 9]
```diff
@@ -74,35 +72,26 @@ public void run() {
...
-        executionTimeStats = new BasicTimer(monitorConfigBuilder.build());
-        cleanupFailed = new BasicCounter(MonitorConfig.builder("Eureka-Connection-Cleaner-Failure").build());
-        try {
-            Monitors.registerObject(this);
-        } catch (Exception e) {
-            logger.error("Unable to register with servo.", e);
-        }
+      final CompositeRegistry registry = Spectator.globalRegistry();
+      executionTimeStats = registry.timer("Eureka-Connection-Cleaner-Time");
+        cleanupFailed = registry.counter("Eureka-Connection-Cleaner-Failure");
```
Comment (brharrington) [inline eureka-client/src/main/java/com/netflix/discovery/shared/transport/jersey/ApacheHttpClientConnectionCleaner.java:77]: indentation

File: eureka-client-jersey2/build.gradle [hunk 10]
```diff
@@ -27,6 +27,8 @@ dependencies {
...
+    compile 'com.netflix.spectator:spectator-api:1.7.3'
```
[the remark this answers is not in the document]
Comment (joshgord) [reply eureka-client-jersey2/build.gradle:31]: Seems like this sub project didn't adopt the version substitution like the others 🤷‍♂️

File: eureka-client-jersey2/src/main/java/com/netflix/discovery/shared/transport/jersey2/EurekaJersey2ClientImpl.java [hunk 11]
```diff
@@ -325,24 +324,19 @@ private PoolingHttpClientConnectionManager createCustomSslCM() {
...
             this.connectionIdleTimeout = connectionIdleTimeout;
-            MonitorConfig.Builder monitorConfigBuilder = MonitorConfig.builder("Eureka-Connection-Cleaner-Time");
-            executionTimeStats = new BasicTimer(monitorConfigBuilder.build());
-            cleanupFailed = new BasicCounter(MonitorConfig.builder("Eureka-Connection-Cleaner-Failure").build());
-            try {
-                Monitors.registerObject(this);
-            } catch (Exception e) {
-                s_logger.error("Unable to register with servo.", e);
-            }
+            final CompositeRegistry registry = Spectator.globalRegistry();
```
[the remark this answers is not in the document]
Comment (joshgord) [reply eureka-client-jersey2/src/main/java/com/netflix/discovery/shared/transport/jersey2/EurekaJersey2ClientImpl.java:332]: Ah right, this was my IDE autocompleting the type, will update

[the remark this answers is not in the document]
Comment (joshgord) [reply eureka-client-jersey2/src/main/java/com/netflix/discovery/shared/transport/jersey2/EurekaJersey2ClientImpl.java]: fixed

File: eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java [hunk 12]
```diff
@@ -362,12 +363,19 @@ public synchronized BackupRegistry get() {
         } else {
             this.registryStalenessMonitor = ThresholdLevelsMetric.NO_OP_METRIC;
         }
+        final CompositeRegistry registry = globalRegistry();
+        PolledMeter.using(registry)
+            .withName(METRIC_REGISTRY_PREFIX + "lastSuccessfulRegistryFetchTimePeriod")
+            .monitorValue(getLastSuccessfulRegistryFetchTimePeriodInternal());
```
[the remark this answers is not in the document]
Comment (joshgord) [reply eureka-client/src/main/java/com/netflix/discovery/DiscoveryClient.java]: Fixed

File: eureka-client-jersey2/build.gradle (hunk 10 above)
[the remark this answers is not in the document]
Comment (brharrington) [reply eureka-client-jersey2/build.gradle:31]: May just need to use double quotes

**Review by brharrington — APPROVED** [review]

**Pull request merged**